File: frontend/apps/ios/Pantopus/Features/Shared/ContentDetail/ContentDetailShell.swift
//
//  ContentDetailShell.swift
//  Pantopus
//
//  Reusable detail scaffold. Concrete screens supply three slots — a
//  header, a body, and an optional CTA — and the shell handles the top
//  bar, scroll container, and back-chevron chrome.
//

import SwiftUI

/// Top-bar trailing action payload for the content-detail shell.
public struct ContentDetailTopBarAction: Sendable {
    public let icon: PantopusIcon
    public let accessibilityLabel: String
    public let handler: @Sendable () -> Void

    public init(icon: PantopusIcon, accessibilityLabel: String, handler: @escaping @Sendable () -> Void) {
        self.icon = icon
        self.accessibilityLabel = accessibilityLabel
        self.handler = handler
    }
}

/// Generic content-detail shell with pluggable header / body / cta slots.
///
/// Rather than wiring three type parameters through the view hierarchy
/// (`Shell<H, B, C>`), we keep the shell ergonomic by taking three
/// `@ViewBuilder` closures — each screen inlines whichever concrete
/// header / body / CTA it uses.
@MainActor
public struct ContentDetailShell<HeaderView: View, BodyView: View, CTAView: View>: View {
    private let title: String?
    private let onBack: (() -> Void)?
    private let topBarAction: ContentDetailTopBarAction?
    private let topBarSecondaryAction: ContentDetailTopBarAction?
    private let onRefresh: (@Sendable () async -> Void)?
    private let headerContent: HeaderView
    private let bodyContent: BodyView
    private let ctaContent: CTAView

    public init(
        title: String? = nil,
        onBack: (() -> Void)? = nil,
        topBarAction: ContentDetailTopBarAction? = nil,
        topBarSecondaryAction: ContentDetailTopBarAction? = nil,
        onRefresh: (@Sendable () async -> Void)? = nil,
        @ViewBuilder header: () -> HeaderView,
        @ViewBuilder body: () -> BodyView,
        @ViewBuilder cta: () -> CTAView = { EmptyView() }
    ) {
        self.title = title
        self.onBack = onBack
        self.topBarAction = topBarAction
        self.topBarSecondaryAction = topBarSecondaryAction
        self.onRefresh = onRefresh
        headerContent = header()
        bodyContent = body()
        ctaContent = cta()
    }

    public var body: some View {
        ZStack(alignment: .bottomTrailing) {
            VStack(spacing: Spacing.s0) {
                ContentDetailTopBar(
                    title: title,
                    onBack: onBack,
                    action: topBarAction,
                    secondaryAction: topBarSecondaryAction
                )
                scroll
                    .background(Theme.Color.appBg)
            }
            ctaContent
                .padding(Spacing.s4)
        }
        .background(Theme.Color.appBg)
    }

    @ViewBuilder private var scroll: some View {
        if let onRefresh {
            scrollContent.refreshable { await onRefresh() }
        } else {
            scrollContent
        }
    }

    private var scrollContent: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: Spacing.s4) {
                headerContent
                bodyContent
                Spacer(minLength: Spacing.s10)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.vertical, Spacing.s4)
        }
    }
}

/// 44pt top bar used by the content-detail shell.
public struct ContentDetailTopBar: View {
    let title: String?
    let onBack: (() -> Void)?
    let action: ContentDetailTopBarAction?
    var secondaryAction: ContentDetailTopBarAction?

    public init(
        title: String?,
        onBack: (() -> Void)?,
        action: ContentDetailTopBarAction?,
        secondaryAction: ContentDetailTopBarAction? = nil
    ) {
        self.title = title
        self.onBack = onBack
        self.action = action
        self.secondaryAction = secondaryAction
    }

    public var body: some View {
        HStack(spacing: Spacing.s2) {
            if let onBack {
                Button(action: onBack) {
                    Icon(.chevronLeft, size: 22, color: Theme.Color.appText)
                        .frame(width: 44, height: 44)
                }
                .buttonStyle(.plain)
                .accessibilityLabel("Back")
            } else {
                Spacer().frame(width: 44, height: 44)
            }
            if secondaryAction != nil {
                // Balance the second trailing button so the title stays centered.
                Spacer().frame(width: 44, height: 44)
            }
            Spacer()
            if let title {
                Text(title)
                    .pantopusTextStyle(.h3)
                    .foregroundStyle(Theme.Color.appText)
                    .lineLimit(1)
                    .accessibilityAddTraits(.isHeader)
            }
            Spacer()
            if let secondaryAction {
                Button(action: secondaryAction.handler) {
                    Icon(secondaryAction.icon, size: 22, color: Theme.Color.appText)
                        .frame(width: 44, height: 44)
                }
                .buttonStyle(.plain)
                .accessibilityLabel(secondaryAction.accessibilityLabel)
            }
            if let action {
                Button(action: action.handler) {
                    Icon(action.icon, size: 22, color: Theme.Color.appText)
                        .frame(width: 44, height: 44)
                }
                .buttonStyle(.plain)
                .accessibilityLabel(action.accessibilityLabel)
            } else {
                Spacer().frame(width: 44, height: 44)
            }
        }
        .padding(.horizontal, Spacing.s2)
        .frame(height: 44)
        .background(Theme.Color.appSurface)
        .overlay(alignment: .bottom) {
            Rectangle().fill(Theme.Color.appBorderSubtle).frame(height: 1)
        }
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/shared/content_detail/ContentDetailShell.kt
@file:Suppress("MagicNumber", "UnusedPrivateMember", "PackageNaming", "LongMethod", "LongParameterList", "MatchingDeclarationName")

package app.pantopus.android.ui.screens.shared.content_detail

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.sizeIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.PantopusTextStyle
import app.pantopus.android.ui.theme.Spacing

/** Trailing top-bar action payload. */
data class ContentDetailTopBarAction(
    val icon: PantopusIcon,
    val contentDescription: String,
    val onClick: () -> Unit,
)

/** Test tag on the shell root. */
const val CONTENT_DETAIL_TAG = "contentDetail"

/**
 * Reusable content-detail scaffold. Concrete screens provide three slots
 * via `@Composable` parameters: `header`, `body`, and an optional `cta`.
 *
 * The shell handles the 44dp top bar (back chevron + centered title + optional
 * action) and the scroll container. No tab bar — this archetype always
 * sits on a nav stack.
 */
@Composable
fun ContentDetailShell(
    title: String?,
    onBack: (() -> Void)? = null,
    topBarAction: ContentDetailTopBarAction? = null,
    topBarSecondaryAction: ContentDetailTopBarAction? = null,
    cta: @Composable () -> Unit = {},
    header: @Composable () -> Unit,
    body: @Composable () -> Unit,
) {
    Box(
        modifier =
            Modifier
                .fillMaxSize()
                .background(PantopusColors.appBg)
                .testTag(CONTENT_DETAIL_TAG),
    ) {
        Column(modifier = Modifier.fillMaxSize()) {
            ContentDetailTopBar(
                title = title,
                onBack = onBack,
                action = topBarAction,
                secondaryAction = topBarSecondaryAction,
            )
            Column(
                modifier =
                    Modifier
                        .fillMaxSize()
                        .verticalScroll(rememberScrollState())
                        .padding(vertical = Spacing.s4),
                verticalArrangement = Arrangement.spacedBy(Spacing.s4),
            ) {
                header()
                body()
                Spacer(Modifier.height(Spacing.s10))
            }
        }
        Box(
            modifier = Modifier.align(Alignment.BottomEnd).padding(Spacing.s4),
        ) {
            cta()
        }
    }
}

/** 44dp top bar. */
@Composable
fun ContentDetailTopBar(
    title: String?,
    onBack: (() -> Void)? = null,
    action: ContentDetailTopBarAction? = null,
    secondaryAction: ContentDetailTopBarAction? = null,
) {
    Column {
        Row(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .background(PantopusColors.appSurface)
                    .height(44.dp)
                    .padding(horizontal = Spacing.s2),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (onBack != null) {
                Box(
                    modifier =
                        Modifier
                            .size(44.dp)
                            .clickable(onClick = onBack)
                            .semantics { contentDescription = "Back" },
                    contentAlignment = Alignment.Center,
                ) {
                    PantopusIconImage(
                        icon = PantopusIcon.ChevronLeft,
                        contentDescription = null,
                        size = 22.dp,
                        tint = PantopusColors.appText,
                    )
                }
            } else {
                Spacer(Modifier.size(44.dp))
            }
            // Balance the trailing pair so the title stays centered.
            if (secondaryAction != null) {
                Spacer(Modifier.size(44.dp))
            }
            Spacer(Modifier.weight(1f))
            if (title != null) {
                Text(
                    text = title,
                    style = PantopusTextStyle.h3,
                    color = PantopusColors.appText,
                    maxLines = 1,
                    modifier = Modifier.semantics { heading() },
                )
            }
            Spacer(Modifier.weight(1f))
            if (secondaryAction != null) {
                Box(
                    modifier =
                        Modifier
                            .size(44.dp)
                            .clickable(onClick = secondaryAction.onClick)
                            .sizeIn(minWidth = 44.dp, minHeight = 44.dp)
                            .semantics { contentDescription = secondaryAction.contentDescription },
                    contentAlignment = Alignment.Center,
                ) {
                    PantopusIconImage(
                        icon = secondaryAction.icon,
                        contentDescription = null,
                        size = 22.dp,
                        tint = PantopusColors.appText,
                    )
                }
            }
            if (action != null) {
                Box(
                    modifier =
                        Modifier
                            .size(44.dp)
                            .clickable(onClick = action.onClick)
                            .sizeIn(minWidth = 44.dp, minHeight = 44.dp)
                            .semantics { contentDescription = action.contentDescription },
                    contentAlignment = Alignment.Center,
                ) {
                    PantopusIconImage(
                        icon = action.icon,
                        contentDescription = null,
                        size = 22.dp,
                        tint = PantopusColors.appText,
                    )
                }
            } else {
                Spacer(Modifier.size(44.dp))
            }
        }
        HorizontalDivider(color = PantopusColors.appBorderSubtle, thickness = 1.dp)
    }
}

// Kotlin alias so the Swift-style `Void` closure type above reads cleanly.
private typealias Void = Unit
